src/core/ddsc/src/shm_monitor.c

```diff
@@ -80,96 +67,55 @@ dds_return_t shm_monitor_wake_and_enable(shm_monitor_t* monitor)
-  struct ddsi_domaingv* gv = rd->m_rd->e.gv;
-  thread_state_awake(lookup_thread_state(), gv);
-
-  while (true)
-  {
-    shm_lock_iox_sub(rd->m_iox_sub);
-    enum iox_ChunkReceiveResult take_result = iox_sub_take_chunk(rd->m_iox_sub, (const void** const)&chunk);
-    shm_unlock_iox_sub(rd->m_iox_sub);
-
-    if (ChunkReceiveResult_SUCCESS != take_result)
-      break;
-   
-    const iceoryx_header_t* ice_hdr = iceoryx_header_from_chunk(chunk);
-
-    // Get writer or proxy writer
-    struct entity_common * e = entidx_lookup_guid_untyped (gv->entity_index, &ice_hdr->guid);
-    if (e == NULL || (e->kind != EK_PROXY_WRITER && e->kind != EK_WRITER))
-    {
-      // Ignore that doesn't match a known writer or proxy writer
-      DDS_CLOG (DDS_LC_SHM, &gv->logconfig, "unknown source entity, ignore.\n");
-      continue;
-    }
-
-    // Create struct ddsi_serdata    
-    struct ddsi_serdata* d = ddsi_serdata_from_iox(rd->m_topic->m_stype, ice_hdr->data_kind, &rd->m_iox_sub, chunk);    
-    d->timestamp.v = ice_hdr->tstamp;
-    d->statusinfo = ice_hdr->statusinfo;
-
-    // Get struct ddsi_tkmap_instance
-    struct ddsi_tkmap_instance* tk;
-    if ((tk = ddsi_tkmap_lookup_instance_ref(gv->m_tkmap, d)) == NULL)
-    {
-      DDS_CLOG(DDS_LC_SHM, &gv->logconfig, "ddsi_tkmap_lookup_instance_ref failed.\n");
-      goto release;
-    }
-
-    // Generate writer_info
-    struct ddsi_writer_info wrinfo;
-    struct dds_qos *xqos;
-    if (e->kind == EK_PROXY_WRITER)
-      xqos = ((struct proxy_writer *) e)->c.xqos;
-    else
-      xqos = ((struct writer *) e)->xqos;
-    ddsi_make_writer_info(&wrinfo, e, xqos, d->statusinfo);
-    (void)ddsi_rhc_store(rd->m_rd->rhc, &wrinfo, d, tk);
-
-release:
-    if (tk)
-      ddsi_tkmap_instance_unref(gv->m_tkmap, tk);
-    if (d)
-      ddsi_serdata_unref(d);
-  }
-  thread_state_asleep(lookup_thread_state());
-}
-
-static void shm_wakeup_trigger_callback(iox_user_trigger_t trigger) 
-{    
-    // we know it is actually in extended storage since we created it like this
-    iox_user_trigger_storage_extension_t* storage = (iox_user_trigger_storage_extension_t*) trigger;
-    if(storage->monitor->m_state == SHM_MONITOR_RUNNING && storage->call) {
-        storage->call(storage->arg);
-    }
+  dds_transfer_samples_from_iox_to_rhc(rd);
 }
 
-static void shm_subscriber_callback(iox_sub_t subscriber) 
+static void shm_subscriber_callback(iox_sub_t subscriber, void * context_data)
 {
+    (void)subscriber;
     // we know it is actually in extended storage since we created it like this
-    iox_sub_storage_extension_t *storage = (iox_sub_storage_extension_t*) subscriber; 
-    if(storage->monitor->m_state == SHM_MONITOR_RUNNING) {
-        receive_data_wakeup_handler(storage->parent_reader);
+    iox_sub_context_t *context = (iox_sub_context_t*) context_data;
+    if((context->monitor) && (context->monitor->m_state == SHM_MONITOR_RUNNING)) {
+        receive_data_wakeup_handler(context->parent_reader);
```

Comment:
We can call `dds_transfer_samples_from_iox_to_rhc(context->parent_reader)` here.